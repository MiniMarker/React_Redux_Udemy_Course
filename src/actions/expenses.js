import uuid from "uuid";
import database from "../firebase/firebase";

// ADD_EXPENSE
export const addExpense = (expense) => ({
	type: "ADD_EXPENSE",
	expense
});

export const startAddExpense = (expenseData = {}) => {

	return (dispatch) => {

		const {
			description = "",
			note = "",
			amount = 0,
			createdAt = 0
		} = expenseData;

		const expense = {description, note, amount, createdAt};

		return database.ref('expenses').push(expense)
			.then((ref) => {

				dispatch(addExpense({
					...expense,
					id: ref.key
				}))

			})
			.catch((err) => {
				console.error(err);
			})
	}
};

// REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({
	type: "REMOVE_EXPENSE",
	id
});

export const startRemoveExpense = (expense) => {
	return (dispatch) => {
		return database.ref(`expenses/${expense.id}`)
			.set(null)
			.then(() => {
				dispatch(removeExpense({
					id: expense.id
				}))
			})
			.catch((err) => {
				console.error("ERROR", err);
			});
	}
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
	type: "EDIT_EXPENSE",
	id,
	updates
});

export const startEditExpense = (id, updates) => {
	return (dispatch) => {
		return database.ref(`expenses/${id}`)
			.set({
				...updates
			})
			.then(() => {
				dispatch(editExpense(id, updates))
			})
			.catch((err) => {
				console.error(err);
			})
	}
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
	type: 'SET_EXPENSE',
	expenses
});

export const startSetExpenses = () => {
	return (dispatch) => {

		return database.ref('expenses')
			.once('value')
			.then((snapshot) => {

				const expenses = [];

				snapshot.forEach((expense) => {
					expenses.push({
						...expense.val(),
						id: expense.key
					});
				});

				dispatch(setExpenses(expenses))

			})
			.catch((err) => {
				console.log("ERROR", err);
			});
	}
};