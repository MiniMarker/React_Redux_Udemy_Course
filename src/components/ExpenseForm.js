import React from "react";
import moment from "moment";
import {SingleDatePicker} from "react-dates";

//const date = new Date();
// const now = moment();
// console.log(now.format("MMM Do, YYYY"));

export default class ExpenseForm extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			errorMessage: '',
			buttonText: props.expense ? "Apply" : "Create"
		};
	}

	onDescriptionChange = (e) => {
		const input = e.target.value;
		this.setState(() => ({
			description: input
		}));
	};

	onNoteChange = (e) => {
		const input = e.target.value;
		this.setState(() => ({
			note: input
		}));
	};

	onAmountChange = (e) => {
		const amount = e.target.value;
		const regex = /^\d{1,}(\.\d{0,2})?$/;

		//use regex.test(<value>) instead of amount.match(<regex>)
		if (!amount || regex.test(amount) ) {
			this.setState(() => ({
				amount
			}));
		}
	};

	/*
	*   DATE!!
	* */

	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({
				createdAt
			}));
		}
	};

	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }));
	};

	/*
	*   SUBMIT
	*/
	onSubmit = (e) => {
		e.preventDefault();

		if (!this.state.description || !this.state.amount) {

			this.setState(() => ({
				errorMessage: 'Please provide description and amount.'
			}));

		} else {

			this.setState(() => ({
				errorMessage: ''
			}));

			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			});
		}
	};

	render() {
		return (
			<div>
				{this.state.errorMessage && <p>{this.state.errorMessage}</p>}
				<form onSubmit={this.onSubmit}>
					<input
						type={"text"}
						placeholder={"Description"}
						autoFocus={true}
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>

					<input
						type={"text"}
						placeholder={"Amount"}
						value={this.state.amount}
						onChange={this.onAmountChange}
					/>

					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
					/>

					<textarea
						placeholder={"Add a note for your expense (optional)"}
						value={this.state.note}
						onChange={this.onNoteChange}
					/>

					<button id={"submit-btn"} type={"submit"}>{this.state.buttonText}</button>
				</form>
			</div>
		);
	}
}

//make the change
