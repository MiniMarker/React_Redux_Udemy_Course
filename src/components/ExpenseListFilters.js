import React from "react";
import {connect} from "react-redux";
import {DateRangePicker} from "react-dates";
import {
	setEndDate,
	setStartDate,
	setTextFilter,
	sortByAmount,
	sortByDate
} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null
	};

	onDatesChange = ({startDate: newStartDate, endDate: newEndDate}) => {
		this.props.setStartDate(newStartDate);
		this.props.setEndDate(newEndDate);
	};

	onFocusChange = (calendarFocused) => {
		this.setState(() => ({
			calendarFocused
		}));
	};

	onTextChange = (e) => {
		this.props.setTextFilter(e.target.value);
	};

	onSortChange = (e) => {
		e.target.value === "date"
			? this.props.sortByDate()
			: this.props.sortByAmount();
	};

	render() {
		return (
			<div>
				<input
					type={"text"}
					value={this.props.filters.text}
					onChange={this.onTextChange}
				/>

				<select
					value={this.props.filters.sortBy}
					onChange={this.onSortChange}
				>
					<option value={"date"}>Date</option>
					<option value={"amount"}>Amount</option>
				</select>

				<DateRangePicker
					startDate={this.props.filters.startDate}
					endDate={this.props.filters.endDate}
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={() => false}
					showClearDates={true}
				/>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => {
	return {
		setTextFilter: (text) => dispatch(setTextFilter(text)),
		sortByDate: () => dispatch(sortByDate()),
		sortByAmount: () => dispatch(sortByAmount()),
		setStartDate: (startDate) => dispatch(setStartDate(startDate)),
		setEndDate: (endDate) => dispatch(setEndDate(endDate))
	}
};

// To access state i need to map the state and add it as a param to connect
const mapStateToProps = (state) => {
	return {
		filters: state.filters
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
