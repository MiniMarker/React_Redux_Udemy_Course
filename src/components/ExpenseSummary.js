import React from "react";
import {connect} from "react-redux";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";
import numeral from "numeral";

export const ExpenseSummary = ({expensesCount, totalAmount}) => (

	<div>
		<p>
			{
				`Showing ${expensesCount} ${expensesCount === 1 ? "expense" : "expenses"} totalling ${numeral(totalAmount / 100).format("$0,0.00")}`
			}
		</p>
	</div>
);

const mapStateToProps = (state) => {

	const visibleExpenses = selectExpenses(state.expenses, state.filters);

	return {
		expensesCount: visibleExpenses.length,
		totalAmount: selectExpensesTotal(visibleExpenses)
	};
};

export default connect(mapStateToProps)(ExpenseSummary);