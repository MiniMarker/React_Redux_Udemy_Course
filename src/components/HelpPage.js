import React from "react";

const HelpPage = () => (
	<div>
		<h1>HelpPage</h1>
		<p>This is the HelpPageComponent</p>
	</div>
);

export default HelpPage;
