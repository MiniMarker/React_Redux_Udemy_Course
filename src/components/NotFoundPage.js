import {Link} from "react-router-dom";
import React from "react";

const NotFoundPage = () => (
	<div>
		<p>404</p>
		<Link to={"/"}>Go home</Link>

	</div>
);

export default NotFoundPage;