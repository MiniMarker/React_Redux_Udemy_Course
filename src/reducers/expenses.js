
// Expenses reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
	switch(action.type) {

		case "ADD_EXPENSE":
			return [
				...state,
				action.expense
			];

		case "REMOVE_EXPENSE":

			//use filter because it returns the updated array, not changing the existing
			return state.filter(( {id} ) => id !== action.id);

		case "EDIT_EXPENSE":
			return state.map((expense) => {
				if(expense.id === action.id) {

					return {
						...expense,
						...action.updates
					}

				} else {
					return expense
				}
			});

		case "SET_EXPENSE":
			return action.expenses;

		default:
			return state;

	}
};