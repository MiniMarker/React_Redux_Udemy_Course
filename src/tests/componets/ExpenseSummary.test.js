import React from 'react';
import {ExpenseSummary} from '../../components/ExpenseSummary';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

test("RenderExpenseSummaryOneValue", () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={1} totalAmount={2312}/>);
	expect(wrapper).toMatchSnapshot();
});

test("RenderExpenseSummaryMultipleValues", () => {
	const wrapper = shallow(<ExpenseSummary expensesCount={expenses.length} totalAmount={2316544672}/>);
	expect(wrapper).toMatchSnapshot();
});