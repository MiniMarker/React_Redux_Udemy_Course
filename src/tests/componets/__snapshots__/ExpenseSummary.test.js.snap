// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderExpenseSummaryMultipleValues 1`] = `
<div>
  <p>
    Showing 3 expenses totalling $23,165,446.72
  </p>
</div>
`;

exports[`RenderExpenseSummaryOneValue 1`] = `
<div>
  <p>
    Showing 1 expense totalling $23.12
  </p>
</div>
`;
